File: python/open_guardrail/guards/encoding_normalize.py
"""Encoding normalization and mixed-encoding attack detection."""
from __future__ import annotations

import re
import time

from open_guardrail.core import GuardResult

_FULLWIDTH = re.compile(r"[\uff01-\uff5e]")
_COMBINING = re.compile(r"[\u0300-\u036f]{2,}")
_MIXED_SCRIPT = re.compile(
    r"[\u0400-\u04ff].*[a-zA-Z]"
    r"|[a-zA-Z].*[\u0400-\u04ff]"
)
_HTML_ENTITY = re.compile(
    r"&(?:#\d{2,5}|#x[0-9a-fA-F]{2,4}|[a-z]+);"
)


class _EncodingNormalize:
    def __init__(self, *, action: str = "block") -> None:
        self.name = "encoding-normalize"
        self.action = action

    def check(
        self, text: str, stage: str = "input"
    ) -> GuardResult:
        start = time.perf_counter()
        issues: list[str] = []

        if _FULLWIDTH.search(text):
            issues.append("fullwidth-chars")
        if _COMBINING.search(text):
            issues.append("combining-diacriticals")
        if _MIXED_SCRIPT.search(text):
            issues.append("mixed-script")
        if _HTML_ENTITY.search(text):
            issues.append("html-entities")

        triggered = len(issues) > 0
        elapsed = (time.perf_counter() - start) * 1000
        return GuardResult(
            guard_name="encoding-normalize",
            passed=not triggered,
            action=self.action if triggered else "allow",
            message=(
                "Encoding anomalies detected"
                if triggered
                else None
            ),
            latency_ms=round(elapsed, 2),
            details=(
                {"issues": issues} if triggered else None
            ),
        )


def encoding_normalize(
    *, action: str = "block"
) -> _EncodingNormalize:
    return _EncodingNormalize(action=action)

File: packages/guards/src/encoding-normalize.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface EncodingNormalizeOptions {
  action: 'block' | 'warn';
}

const FULLWIDTH_RE = /[\uFF01-\uFF5E]/;
const COMBINING_RE = /[\u0300-\u036F]{2,}/;
const MIXED_SCRIPT_RE = /[\u0400-\u04FF].*[a-zA-Z]|[a-zA-Z].*[\u0400-\u04FF]/;
const HTML_ENTITY_RE = /&(?:#\d{2,5}|#x[0-9a-fA-F]{2,4}|[a-z]+);/;

export function encodingNormalize(options: EncodingNormalizeOptions): Guard {
  return {
    name: 'encoding-normalize',
    version: '0.1.0',
    description: 'Detects mixed encoding attacks: fullwidth, combining chars, mixed scripts, HTML entities',
    category: 'security',
    supportedStages: ['input'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const issues: string[] = [];

      if (FULLWIDTH_RE.test(text)) issues.push('fullwidth-chars');
      if (COMBINING_RE.test(text)) issues.push('combining-diacriticals');
      if (MIXED_SCRIPT_RE.test(text)) issues.push('mixed-script');
      if (HTML_ENTITY_RE.test(text)) issues.push('html-entities');

      const triggered = issues.length > 0;

      return {
        guardName: 'encoding-normalize',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        latencyMs: Math.round(performance.now() - start),
        details: triggered ? { issues, reason: 'Encoding anomalies detected' } : undefined,
      };
    },
  };
}
